include/cetl/variable_length_array.hpp

```diff
@@ -0,0 +1,2065 @@
+    /// Required constructor for std::uses_allocator protocol.
+    /// @param alloc Allocator to use for all allocations.
+    explicit VariableLengthArray(const allocator_type& alloc) noexcept
+        : Base(alloc, nullptr, 0, 0, std::numeric_limits<size_type>::max())
+    {
+    }
+
+    /// Required constructor for std::uses_allocator protocol.
+    /// @param alloc Allocator to use for all allocations.
+    /// @param max_size_max Clamping value for the maximum size of this array. That is,
+    /// cetl::VariableLengthArray::max_size() will return
+    /// `std::min(max_size_max, std::allocator_traits<allocator_type>::max_size(alloc))`
+    explicit VariableLengthArray(const allocator_type& alloc, size_type max_size_max) noexcept
+        : Base(alloc, nullptr, 0, 0, max_size_max)
+    {
+    }
+
+    /// Initializer syntax constructor.
+    /// @param l Initializer list of elements to copy into the array.
+    /// @param alloc Allocator to use for all allocations.
+    /// @param max_size_max Clamping value for the maximum size of this array. That is,
+    /// cetl::VariableLengthArray::max_size() will return
+    /// `std::min(max_size_max, std::allocator_traits<allocator_type>::max_size(alloc))`
+    VariableLengthArray(std::initializer_list<value_type> l,
+                        const allocator_type&             alloc,
+                        size_type                         max_size_max = std::numeric_limits<size_type>::max())
+        : Base(alloc, nullptr, 0, 0, max_size_max)
+    {
+        reserve(l.size());
+        size_ = Base::fast_copy_construct(data_, capacity_, l.begin(), l.size(), alloc_);
+    }
+
+    /// Range constructor.
+    /// @tparam InputIt The type of the range's iterators.
+    /// @param first    The beginning of the range.
+    /// @param last     The end of the range.
+    /// @param alloc    Allocator to use for all allocations.
+    /// @param max_size_max Clamping value for the maximum size of this array. That is,
+    /// cetl::VariableLengthArray::max_size() will return `std::min(max_size_max,
+    /// std::allocator_traits<allocator_type>::max_size(alloc))`
+    template <class InputIt>
+    VariableLengthArray(InputIt               first,
+                        InputIt               last,
+                        const allocator_type& alloc,
+                        size_type             max_size_max = std::numeric_limits<size_type>::max())
+        : Base(alloc, nullptr, 0, 0, max_size_max)
+    {
+        if (last >= first)
+        {
+            reserve(static_cast<std::size_t>(last - first));
+            if (nullptr != data_)
+            {
+                size_ =
+                    Base::fast_copy_construct(data_, capacity_, first, static_cast<std::size_t>(last - first), alloc_);
+            }
+        }
+    }
+
+    //
+    // Rule of Five.
+    //
+    VariableLengthArray(const VariableLengthArray& rhs)
+        : Base(rhs)
+    {
+        Base::reserve(rhs.size(), rhs.max_size());
```

Comment:
Copy construction should preserve the capacity.
I want the following to work.
```
std::array<std::byte, 2 * 10 * sizeof(std::uint32)> buffer{};
cetl::pf17::pmr::monotonic_buffer_resource mbr{buffer};
cetl::pf17::pmr::polymorphic_allocator<std::uint32_t> pa{&mbr};

cetl::VariableLengthArray<std::uint32_t> foo{pa, 10};
foo.reserve(10);
for (auto i = 0; i < 5; i++) { foo.push_back(i); }

cetl::VariableLengthArray<std::uint32_t> foo2{foo};
// I've already planned my buffer and know that two instances of this VLA
// will fit.  I don't want this next statement to allocate.  I want foo2
// to have the same capacity as foo.
foo2.push_back(5);
```

Reply:
Looked into this and I don't think this is the correct default behaviour so I added #30 to follow up.